### PR TITLE
malloc: add checked deallocator

### PR DESCRIPTION
**Pull request by reusee**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #16059 

## What this PR does / why we need it:
malloc: add TestCheckedDeallocator

malloc: more logs

add code owner

### CODE REVIEW

**Comment by matrix-meow** [comment]
@reusee Thanks for your contributions!

### Pull Request Review:

1. **Title and Body:**
   - The title of the pull request is clear and concise, indicating the addition of a checked deallocator for malloc.
   - The body of the pull request provides information about the type of PR, the related issue, and the purpose of the changes. It is well-structured and informative.

2. **Changes Made:**
   - **Added Checked Deallocator:**
     - A new file `checked_deallocator.go` and its corresponding test file have been added to the `pkg/common/malloc` directory.
     - The `checkedDeallocator` struct and its methods have been implemented to handle deallocation with checks for memory leaks and double frees.
     - The `newCheckedDeallocator` function sets a finalizer to check for memory leaks upon deallocation.
     - The `Deallocate` method in `checkedDeallocator` ensures that double frees are detected and handled appropriately.
   - **Modified Class Allocator:**
     - Changes have been made in `class_allocator.go` to support a new parameter `checkDeallocate` in `NewClassAllocator`.
     - The `fixedSizeMmapAllocator` now includes a flag `checkDeallocate` to enable checked deallocation.
     - The `Allocate` method in `fixedSizeMmapAllocator` now conditionally creates a `checkedDeallocator` based on the `checkDeallocate` flag.
   - **Modified Default Allocator:**
     - Changes in `default.go` to enable the checked deallocator based on the environment variable `MO_MALLOC_CHECKED`.
   - **Test Changes:**
     - Test files `checked_deallocator_test.go`, `class_allocator_test.go`, `data_test.go`, and `rc_bytes_test.go` have been updated to reflect the changes in the allocator setup.

3. **Issues/Improvements:**
   - **Security Concerns:**
     - The addition of a checked deallocator is a good practice to prevent memory leaks and double frees, enhancing memory safety.
     - However, the implementation should be reviewed thoroughly to ensure there are no vulnerabilities or bypasses in the memory checking mechanism.
     - Consider additional security testing, like fuzz testing, to identify any potential issues with the new checked deallocator.
   - **Optimization:**
     - The code changes seem to be focused on adding the checked deallocator functionality, which is a valuable improvement.
     - To optimize the changes, consider refactoring or restructuring the code to improve readability and maintainability.
     - Ensure that the new code

File: pkg/common/malloc/default.go [hunk 1]
```diff
@@ -17,9 +17,28 @@ package malloc
...
+	"go.uber.org/zap"
 )
 
+// checkFraction controls the fraction of checked deallocations
+// On average, 1 / fraction of deallocations will be checked for double free or missing free
+var checkFraction = func() uint32 {
+	// default fraction, large enough to not effecting performance too much
+	fraction := uint32(8192)
+	// control with env var
+	if str := os.Getenv("MO_MALLOC_CHECK_FRACTION"); str != "" {
```
Comment (fengttt) [inline pkg/common/malloc/default.go:33]: Use our config system for this, not os.Getenv

File: pkg/common/malloc/default.go [hunk 2]
```diff
@@ -17,9 +17,28 @@ package malloc
...
 	"strings"
+
+	"github.com/matrixorigin/matrixone/pkg/logutil"
+	"go.uber.org/zap"
 )
 
+// checkFraction controls the fraction of checked deallocations
+// On average, 1 / fraction of deallocations will be checked for double free or missing free
+var checkFraction = func() uint32 {
+	// default fraction, large enough to not effecting performance too much
```
Comment (fengttt) [inline pkg/common/malloc/default.go:30]: default should be 0, that is, no debug alloc.

File: pkg/common/malloc/class_allocator.go [hunk 3]
```diff
@@ -100,16 +106,24 @@ func newFixedSizedAllocator(size uint64) *fixedSizeMmapAllocator {
...
 }
 
 var _ Allocator = new(fixedSizeMmapAllocator)
 
-func (f *fixedSizeMmapAllocator) Allocate(_ uint64) (unsafe.Pointer, Deallocator) {
+func (f *fixedSizeMmapAllocator) Allocate(_ uint64) (ptr unsafe.Pointer, dec Deallocator) {
+	defer func() {
+		n := fastrand()
+		if n%f.checkFraction == 0 {
+			dec = newCheckedDeallocator(dec)
```
Comment (fengttt) [inline pkg/common/malloc/class_allocator.go:123]: use f.checkFraction as 0 to disable this checked allocator/deallocator.

**Review by fengttt — CHANGES_REQUESTED** [review]

File: pkg/common/malloc/default.go (hunk 1 above)
Comment (reusee) [reply pkg/common/malloc/default.go]: done

File: pkg/common/malloc/default.go (hunk 2 above)
Comment (reusee) [reply pkg/common/malloc/default.go]: done

File: pkg/common/malloc/class_allocator.go (hunk 3 above)
Comment (reusee) [reply pkg/common/malloc/class_allocator.go:123]: done

**Review by fengttt — APPROVED** [review]

**Pull request merged**